compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DocumentationAnalyzer.java

```diff
@@ -167,6 +211,117 @@ public void visit(BLangResource resourceNode) {
+            if (!isValidIdentifier) {
+                dlog.warning(reference.pos, DiagnosticCode.INVALID_DOCUMENTATION_REFERENCE,
+                        reference.referenceName, reference.getType().getValue());
+            }
+        });
+    }
+
+    private boolean validateIdentifier(DiagnosticPos pos,
+                                       DocumentableNode documentableNode,
+                                       DocumentationReferenceType type,
+                                       String packageId, String typeID,
+                                       String identifier) {
+        int tag = -1;
+        //Lookup namespace to validate the identifier
+        switch (type) {
+            case PARAMETER:
+                //Parameters are only available for function nodes.
+                if (documentableNode.getKind() == NodeKind.FUNCTION) {
+                    BLangFunction funcNode = (BLangFunction) documentableNode;
+                    SymbolEnv funcEnv = SymbolEnv.createFunctionEnv(funcNode, funcNode.symbol.scope, this.env);
+                    return (resolveFullyQualifiedSymbol(pos, funcEnv, packageId, typeID, identifier, SymTag.VARIABLE)
+                            != symTable.notFoundSymbol);
+                } else {
+                    return false;
+                }
+            case SERVICE:
+                tag = SymTag.SERVICE;
+                break;
+            case TYPE:
+                tag = SymTag.TYPE;
+                break;
+            case VARIABLE:
+            case VAR:
+                tag = SymTag.VARIABLE;
+                break;
+            case ANNOTATION:
+                tag = SymTag.ANNOTATION;
+                break;
+            case MODULE:
+                tag = SymTag.IMPORT;
+                break;
+            case CONST:
+                tag = SymTag.CONSTANT;
+                break;
+            case BACKTICK_CONTENT:
+            case FUNCTION:
+                tag = SymTag.FUNCTION;
+                break;
+        }
+        return (resolveFullyQualifiedSymbol(pos, this.env, packageId, typeID, identifier, tag)
+                != symTable.notFoundSymbol);
+    }
+
+    private BSymbol resolveFullyQualifiedSymbol(DiagnosticPos pos, SymbolEnv env, String packageId, String type,
+                                                String identifier, int tag) {
+        Name identifierName = names.fromString(identifier);
+        Name pkgName = names.fromString(packageId);
+        Name typeName = names.fromString(type);
+        SymbolEnv pkgEnv = env;
+
+        if (pkgName != Names.EMPTY) {
+            BSymbol pkgSymbol = symResolver.resolvePrefixSymbol(env, pkgName,
+                            names.fromString(pos.getSource().getCompilationUnitName()));
+
+            if (pkgSymbol == symTable.notFoundSymbol) {
+                return symTable.notFoundSymbol;
+            }
+
+            if (pkgSymbol instanceof BPackageSymbol) {
```

Comment:
Cant we use `pkgSymbol.tag == SymTag.PACKAGE` here instead of the `instanceof` check?